### DISCUSSION TITLE
[eglot] (warning) Server tried to register unsupported capability `workspace/didChangeWorkspaceFolders'

### DISCUSSION THREAD

**Discussion by LuciusChen** [joaotavora/eglot#1237] [General]
When using Eglot with my LSP server, I encountered the following warning/error message: "server tried to register unsupported capability 'workspace/didChangeWorkspaceFolders'". This appears to be due to the LSP server attempting to register the 'workspace/didChangeWorkspaceFolders' capability, which is not currently supported by Eglot.

This capability allows the server to receive notifications when the folders in the workspace change. While this warning/error does not prevent me from using Eglot for code editing and navigation, it does seem to indicate that some features provided by the LSP server might not be available due to this unsupported capability.

I would like to request that support for 'workspace/didChangeWorkspaceFolders' be considered for a future update of Eglot to fully take advantage of all the features offered by my LSP server. Thank you.

**Comment by joaotavora** [comment]
> This capability allows the server to receive notifications when the folders in the workspace change.

https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_didChangeWorkspaceFolders

What exactly do you do in the client, which in this case is Emacs, that constitutes a change of "workspace folders" according to the definition above?

>  it does seem to indicate that some features provided by the LSP server might not be available due to this unsupported capability.

What features would those be?

**Reply by LuciusChen** [reply]
The warning appears just by opening a .vue file in the project.

**Reply by joaotavora** [reply]
Hmmm, this conversation isn't going far.  When I ask "what exactly do you do in the client", I ask in the sense of the checklist that you can consult in:

https://joaotavora.github.io/eglot/#Troubleshooting-Eglot

Here's it's a feature request you're after but for me to understand what you want to do you still need to supply those steps 1-7, particuarly  steps 4 and 6 in this case.

If, on the other hand, you just want ot get rid of the warning you must ask the server developers (whatever your server is) to not try to register a capability for which Eglot doesn't advertise `:dynamicRegistration` capabilities.

**Comment by simurgh9** [comment]
I have the same issue but with Java files.
- Eglot: `Version: 1.17`
- Emacs: `GNU Emacs 29.4 (build 1, aarch64-apple-darwin21.6.0, NS appkit-2113.60 Version 12.6.6 (Build 21G646)) of 2024-08-02`
- OS: MacOS Sonoma 14.6.1

Steps to reproduce,
1. Run `emacs TestFile.java -Q -f package-initialize -f eglot`.
2. You should see the warning in the mini-buffer area.

The language server I am using is `jdtls: stable 1.38.0`.  Here is the error on my system.

<img src="https://github.com/user-attachments/assets/2bc6bb4e-a515-4bb8-b22b-2da9ff08c481" width="500">

**Reply by joaotavora** [reply]
As I wrote above:

> if you just want ot get rid of the warning you must ask the server developers (whatever your server is) to not try to register a capability for which Eglot doesn't advertise :dynamicRegistration capabilities.

If you want me to support this capability , then help me understand

> what exactly do you do in the client, which in this case is Emacs, that constitutes a change of "workspace folders" according to the definition above?

**Reply by simurgh9** [reply]
I clearly do not understand the lsp particulars as you do. I just know it as the system that when enabled gives me autocomplete. My response was more so replying to,

> Here's it's a feature request you're after but for me to understand what you want to do you still need to supply those steps 1-7, particuarly steps 4 and 6 in this case.

Since it give you a minimal way to reproduce the warning with versions and all.

> what exactly do you do in the client, which in this case is Emacs, that constitutes a change of "workspace folders"

Open a java file and launch eglot. Whether or not it follows the definition, well I am not sure.

**Reply by joaotavora** [reply]
> am not sure.

I assure you it does not. 

`didChangeWorkspaceFolders` is, not unsurprisingly, a feature about changing folders and it does not seem you have changed a folder in your recipe. So the feature is useless for you.

**Reply by simurgh9** [reply]
> is, not unsurprisingly, a feature about changing folders and it does not seem you have changed a folder in your recipe. So the feature is useless for you.

Yes, yes I am aware that I did not change the folder.  Why is the warning displayed then?  Are you saying that two different servers happen to have the same bug that is causing this warning and there is absolutely no possibility that this could be an interpretation problem on Eglot's end?

**Reply by joaotavora** [reply]
> Yes, yes I am aware that I did not change the folder. Why is the warning displayed then? Are you saying that two different servers happen to have the same bug that is causing this warning and there is absolutely no possibility that this could be an interpretation problem on Eglot's end?

Yes.

If you have evidence of the contrary, please show it, and thanks in advance.

**Reply by RolKau** [reply]
It seems that all language servers that uses the vscode-languageserver-node has this flaw: https://github.com/microsoft/vscode-languageserver-node/issues/713 (as well as JDTLS)
